crates/apollo-compiler/src/database/hir.rs

```diff
@@ -1974,6 +1998,25 @@ impl ObjectTypeDefinition {
         )
     }
 
+    /// Introspection fields that must not be returned as part of `fields` method.
+    pub(crate) fn implicit_fields(&self, db: &dyn HirDatabase) -> &[FieldDefinition] {
+        let is_root_query = db
+            .schema()
+            .root_operations()
+            .any(|op| op.operation_ty().is_query() && op.named_type().name() == self.name());
+        if is_root_query {
```

Comment:
Is it possible to provide the correct list when we create the HIR node for this, so we dont have to filter it down after the fact?

Reply:
right, so i am not sure how to get this right while we are creating the hir node. running a check to see if this object type is defined as the the root `query` type requires a db query, which causes a cycle. any ideas?

Reply:
Hmm. I think what bothers me a little in this method is that we have this `Arc<ImplicitFields>` list in each node when it isn't really specific to that node, especially since it's also kinda the wrong list of fields in the case of non-root operation types.

what if we didn't store `implicit_fields` in the object types themselves, but had a `db.implicit_fields()` query that returns the appropriate implicit fields for a given object type? The implementation wouldnt need to be very different most likely

Reply:
i've had something like `db.implicit_fields()` before, iirc there were some lifetime issues with that. let me try again!

Reply:
Is it actually a cycle? I’d expect `db.schema()` to be able to complete without constructing HIR for object types

Reply:
yea, it calls `db.object_types` to relate root operations to the existing object types 

https://github.com/apollographql/apollo-rs/blob/aaf8d14e58fe94ee5375391d33bc9ad6bcaa08e3/crates/apollo-compiler/src/database/hir_db.rs#L655

Reply:
> maybe i'll create a private hir type that we can filter afterwards, but that will mean that that hir type will still have an Arc<Vec<FieldDefinition>>

nope this also doesn't work hm hm hmmm